### PR TITLE
[new feature] multi optimMethods support in Optimizer

### LINKED ISSUE #2171: Support multi-OptimMethod in one model.
For model like wide&deep can use more than one OptimMethods in model's different parts. Like wide&deep model,  ftrl for wide part, adagrad for deep part.

### PR DESCRIPTION
**Pull request by qiuxin2012**
## What changes were proposed in this pull request?

multi optimMethods support in Optimizer

Add a new interface def setOptimMethods(method: Map[String, OptimMethod[T]]) to set multi OptimMethods in Optimizer.
With this API, user can set different optimMethods for each subModules with the subModules' name and optimMethod pair.

Updated #2275

## How was this patch tested?

unit test

## Related links or issues (optional)
fixed #2171

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 1]
```diff
@@ -62,7 +65,7 @@ object DistriOptimizer {
     localStates: Array[Table],
     var moduleTimeList: Array[Long] = null,
     localMethods: Array[Option[Array[ValidationMethod[T]]]],
-    optimMethod: OptimMethod[T]
+    optimMethod: Map[String, OptimMethod[T]]
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: optimMethods ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 2]
```diff
@@ -112,24 +118,26 @@ object DistriOptimizer {
...
-    if (!optimMethod.state.contains("epoch")) optimMethod.state.update("epoch", 1)
-    if (!optimMethod.state.contains("neval")) optimMethod.state.update("neval", 1)
-    if (!optimMethod.state.contains("Loss")) {
-      optimMethod.state.update("Loss", Float.PositiveInfinity)
-    }
-    if (!optimMethod.state.contains("score")) optimMethod.state.update("score", 0f)
-    if (!optimMethod.state.contains("recordsProcessedThisEpoch")) {
-      optimMethod.state.update("recordsProcessedThisEpoch", 0)
+    optimMethods.values.foreach{ optimMethod =>
+      if (!optimMethod.state.contains("epoch")) optimMethod.state.update("epoch", 1)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Call optimMethods share states as all seesm to be duplicated here

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 3]
```diff
@@ -409,13 +435,14 @@ object DistriOptimizer {
...
 
-        optimMethod.state.update("recordsProcessedThisEpoch", recordsProcessedThisEpoch)
-
-        optimMethod.state.update("epoch", driverState[Int]("epoch"))
-        optimMethod.state.update("neval", driverState[Int]("neval"))
-        optimMethod.state.update("Loss", driverState[Float]("Loss"))
-        if (validationMethods.isDefined) {
-          optimMethod.state.update("score", driverState[Float]("score"))
+        optimMethods.map { case (moduleName, optimMethod) =>
+          optimMethod.state.update("recordsProcessedThisEpoch", recordsProcessedThisEpoch)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Can optimMethods share the same state ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala [hunk 4]
```diff
@@ -227,22 +236,42 @@ abstract class Optimizer[T: ClassTag, D](
...
+    if (optimMethods.size == 1 && optimMethods.contains(model.getName())) {
+      if (newModel.getName() != model.getName()) {
+        optimMethods = Map(newModel.getName() -> optimMethods(model.getName()))
+      }
+      logger.info(s"Optimizer.setModel: Detect current optimMethod is a global optimMethod." +
+        s" Automatically associate the current optimMethod with the new model.")
+    } else {
+      logger.warn(s"Optimizer.setModel: Detect current optimMethod is not a global optimMethod." +
+        s" Clearing the old optimMethod.")
+      optimMethods = Map[String, OptimMethod[T]]()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Optimizer.scala]: should we build a newModelName -> optimMethod map here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala [hunk 5]
```diff
@@ -87,8 +87,20 @@ class LocalOptimizer[T: ClassTag] (
...
     var wallClockTime = 0L
     var count = 0
-    optimMethod.clearHistory()
-    optimMethod.loadFromTable(state)
+    optimMethods.values.foreach{ optimMethod =>
+      optimMethod.clearHistory()
+    }
+
+    // To be compatible with the old usage that user define hyperparameters in a table.
+    if (optimMethods.size == 1) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/LocalOptimizer.scala]: should we throw exception or print warnings if the size is not 1?

**Comment by yiheng** [comment]
please check my comments

**Comment by wzhongyuan** [comment]
Can we run a full training on a specific model to check the compatibility ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala (hunk 3 above)
Comment (qiuxin2012) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: We can't, the state also saves some internal buffer, the buffer will conflict if we share states.

**Comment by qiuxin2012** [comment]
jenkins passed

**Pull request merged**